man/queue/sp_queue_qremove.3
.\"  Staple - A general-purpose data structure library in pure C89.
.\"  Copyright (C) 2021  Randoragon
.\"
.\"  This library is free software; you can redistribute it and/or
.\"  modify it under the terms of the GNU Lesser General Public
.\"  License as published by the Free Software Foundation;
.\"  version 2.1 of the License.
.\"
.\"  This library is distributed in the hope that it will be useful,
.\"  but WITHOUT ANY WARRANTY; without even the implied warranty of
.\"  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
.\"  Lesser General Public License for more details.
.\"
.\"  You should have received a copy of the GNU Lesser General Public
.\"  License along with this library; if not, write to the Free Software
.\"  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
.\"--------------------------------------------------------------------------------
.TH SP_QUEUE_QREMOVE 3 DATE "libstaple-VERSION"
.SH NAME
sp_queue_qremove,
sp_queue_qremovec,
sp_queue_qremoves,
sp_queue_qremovei,
sp_queue_qremovel,
sp_queue_qremovesc,
sp_queue_qremoveuc,
sp_queue_qremoveus,
sp_queue_qremoveui,
sp_queue_qremoveul,
sp_queue_qremovef,
sp_queue_qremoved,
sp_queue_qremoveld,
sp_queue_qremovell,
sp_queue_qremoveull,
sp_queue_qremoveu8,
sp_queue_qremoveu16,
sp_queue_qremoveu32,
sp_queue_qremoveu64,
sp_queue_qremovei8,
sp_queue_qremovei16,
sp_queue_qremovei32,
sp_queue_qremovei64,
sp_queue_qremovestr
\- quickly remove an element from any position of a queue
.SH SYNOPSIS
.ad l
#include <staple.h>
.sp
int
.BR sp_queue_qremove "(struct sp_queue"
.RI * queue ,
size_t
.IR idx ,
int
.RI (* dtor )(void*))
.br
char
.BR sp_queue_qremovec "(struct sp_queue"
.RI * queue ,
size_t
.IR idx )
.br
short
.BR sp_queue_qremoves "(struct sp_queue"
.RI * queue ,
size_t
.IR idx )
.br
int
.BR sp_queue_qremovei "(struct sp_queue"
.RI * queue ,
size_t
.IR idx )
.br
long
.BR sp_queue_qremovel "(struct sp_queue"
.RI * queue ,
size_t
.IR idx )
.br
signed char
.BR sp_queue_qremovesc "(struct sp_queue"
.RI * queue ,
size_t
.IR idx )
.br
unsigned char
.BR sp_queue_qremoveuc "(struct sp_queue"
.RI * queue ,
size_t
.IR idx )
.br
unsigned short
.BR sp_queue_qremoveus "(struct sp_queue"
.RI * queue ,
size_t
.IR idx )
.br
unsigned int
.BR sp_queue_qremoveui "(struct sp_queue"
.RI * queue ,
size_t
.IR idx )
.br
unsigned long
.BR sp_queue_qremoveul "(struct sp_queue"
.RI * queue ,
size_t
.IR idx )
.br
float
.BR sp_queue_qremovef "(struct sp_queue"
.RI * queue ,
size_t
.IR idx )
.br
double
.BR sp_queue_qremoved "(struct sp_queue"
.RI * queue ,
size_t
.IR idx )
.br
long double
.BR sp_queue_qremoveld "(struct sp_queue"
.RI * queue ,
size_t
.IR idx )
.br
long long
.BR sp_queue_qremovell "(struct sp_queue"
.RI * queue ,
size_t
.IR idx )
.br
unsigned long long
.BR sp_queue_qremoveull "(struct sp_queue"
.RI * queue ,
size_t
.IR idx )
.br
uint8_t
.BR sp_queue_qremoveu8 "(struct sp_queue"
.RI * queue ,
size_t
.IR idx )
.br
uint16_t
.BR sp_queue_qremoveu16 "(struct sp_queue"
.RI * queue ,
size_t
.IR idx )
.br
uint32_t
.BR sp_queue_qremoveu32 "(struct sp_queue"
.RI * queue ,
size_t
.IR idx )
.br
uint64_t
.BR sp_queue_qremoveu64 "(struct sp_queue"
.RI * queue ,
size_t
.IR idx )
.br
int8_t
.BR sp_queue_qremovei8 "(struct sp_queue"
.RI * queue ,
size_t
.IR idx )
.br
int16_t
.BR sp_queue_qremovei16 "(struct sp_queue"
.RI * queue ,
size_t
.IR idx )
.br
int32_t
.BR sp_queue_qremovei32 "(struct sp_queue"
.RI * queue ,
size_t
.IR idx )
.br
int64_t
.BR sp_queue_qremovei64 "(struct sp_queue"
.RI * queue ,
size_t
.IR idx )
.br
char
.RB * sp_queue_qremovestr "(struct sp_queue"
.RI * queue ,
size_t
.IR idx )
.sp
Link with \fI-lstaple\fP.
.ad
.SH DESCRIPTION
See
.BR sp_queue_remove (3).
.P
The functions in the
.BR sp_queue_qremove ()
family operate in constant time, which is achievable by changing the order of
the elements during removal.
.SH CONFORMING TO
See
.BR libstaple (7).
.SH EXAMPLES
To explain how the
.BR sp_queue_remove ()
and
.BR sp_queue_qremove ()
families differ, consider the following initial queue (from head to tail):
.IP
.B 1 2 3 4 _ _ _
.P
Now, let us investigate what happens when we try to remove the number
.BR 2 .
.P
The functions in the
.BR sp_queue_remove ()
family would remove the element, and then move
.BR 3 " and " 4
forward into the freed spot, without affecting the order:
.IP (1)
.B 1 _ 3 4 _ _ _
.br
\h'2n'^
.IP (2)
.B 1 3 4 _ _ _ _
.br
\h'2n'^ ^
.P
On the contrary, the functions in the
.BR sp_queue_qremove ()
family achieve the goal faster by moving the tail of the queue (in this case
.BR 4 )
into the value that should be removed:
.IP (1)
.B 1 _ 3 4 _ _ _
.br
\h'2n'^
.IP (2)
.B 1 4 3 _ _ _ _
.br
\h'2n'^
.P
The second approach guarantees constant time, but does not preserve the order of
the elements.
.SH SEE ALSO
.ad l
.BR libstaple (7),
.BR sp_queue (7),
.BR sp_queue_create (3),
.BR sp_queue_destroy (3),
.BR sp_queue_clear (3),
.BR sp_queue_push (3),
.BR sp_queue_peek (3),
.BR sp_queue_pop (3),
.BR sp_queue_insert (3),
.BR sp_queue_remove (3),
.BR sp_queue_qinsert (3),
.BR sp_queue_get (3),
.BR sp_queue_set (3),
.BR sp_queue_eq (3),
.BR sp_queue_copy (3),
.BR sp_queue_map (3),
.BR sp_queue_print (3)


src/queue/sp_queue_qremove.c
/*  Staple - A general-purpose data structure library in pure C89.
 *  Copyright (C) 2021  Randoragon
 *
 *  This library is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU Lesser General Public
 *  License as published by the Free Software Foundation;
 *  version 2.1 of the License.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *  Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public
 *  License along with this library; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 */
#include "../sp_queue.h"
#include "../internal.h"
#include "../sp_errcodes.h"
#include <string.h>

int sp_queue_qremove(struct sp_queue *queue, size_t idx, int (*dtor)(void*))
{
	char *p;
	int err;
#ifdef STAPLE_DEBUG
	if (queue == NULL) {
		error(("queue is NULL"));
		return SP_EINVAL;
	}
	if (idx >= queue->size) {
		error(("index out of range"));
		return SP_EINDEX;
	}
#endif
	p = sp_ringbuf_get(idx, queue->data, queue->capacity, queue->elem_size, queue->head);
	if (dtor != NULL && (err = dtor(p))) {
		error(("callback function dtor returned %d (non-0)", err));
		return SP_ECALLBK;
	}
	memcpy(p, queue->tail, queue->elem_size);
	if (queue->size != 1)
		sp_ringbuf_decr(&queue->tail, queue->data, queue->capacity, queue->elem_size);
	--queue->size;
	return 0;
}
